src/jsCodegen/CompiledJavaScript.ts

```diff
@@ -0,0 +1,29 @@
+import DomFacade from '../domFacade/DomFacade';
+import Value from '../expressions/dataTypes/Value';
+
+export type CompiledJavaScriptResult =
+	| { isAstAccepted: true; result: CompiledJavaScript }
+	| { isAstAccepted: false; reason: string };
+
+export default class CompiledJavaScript {
+	private compiledJavaScript: string;
+	private fn: any;
+	private runtimeLibrary: any;
+
+	constructor(compiledJavaScript: string, runtimeLibrary: any) {
```

Comment:
I am still not sure whether including this here is the best idea. This does make using the codegen backend very easy, but it also blocks off generating JavaScript files on disk that contain the precompiled XPaths.

The codegen back-end is very much in beta and it will be changed. I'm kind-of fine keeping this as is, but like I said, not sure.

@jarivm what are your thoughts?

Reply:
Also, lets type the runtimeLibrary

Reply:
I've replaced it with this:

```
let compiledJavaScriptFunction = compilationOptions.disableCache
    ? null
    : getCompiledJavaScriptFromCache(expressionString, returnType);

if (!compiledJavaScriptFunction) {
    const ast = parseExpression(expressionString, compilationOptions);
    const compiledJavaScriptResult = compileAstToJavaScript(ast, returnType);

    if (compiledJavaScriptResult.isAstAccepted) {
        // tslint:disable-next-line
        compiledJavaScriptFunction = new Function(
            'contextItem',
            'domFacade',
            'runtimeLibrary',
            compiledJavaScriptResult.code
        );

        storeCompiledJavaScriptInCache(
            expressionString,
            returnType,
            compiledJavaScriptFunction
        );
    }
}

if (compiledJavaScriptFunction) 
    return evaluateCompiledJavaScript(compiledJavaScriptFunction, contextItem, domFacade);
} else {
    // other things
}
```

`evaluateCompiledJavaScript` and `compileAstToJavaScript` could be exposed. That way storing compiled JS is possible. 

We should, however, look for an alternative for the `new Function`. .

Reply:
The easiest way to prevent a `new Function` is I guess to follow your comment and expose a `compileXPathToJavaScript` function that accepts all the static info: namespace bindings, expected return type and the other static things, returning a JavaScript source code string. Plus an `executeCompiledXPath` function that accepts an instance of a function, a node, etcetera and evaluates the function with the runtime lib plus all the arguments.

That way, users can choose to either `new Function` the JS source code string to a function instance, or just output the acquired string to disk to load it later. When storing it on disk, I think it is likely they will be something like ``` const precompiledXPaths = {'self::p': function (runtimeLib, node, domFacade) {<whatever the compiler barfed out>}, 'self::div': <more compiled code>} . That code can be minified, served, parsed by the browser and passed into the evaluate function.

Reply:
Would this work?
```javascript
it.only('compiles XPath to an executable JavaScript function', () => {
  const supportedQuery = '/xml';
  const unsupportedQuery = '/xml[@id = "yes"]';

  // Compile time.
  const precompiledXPaths = {};
  function compileIfPossible(query: string, returnType: ReturnType) {
    const jsResult = compileXPathToJavaScript(query, returnType);
    if (jsResult.isAstAccepted) {
      precompiledXPaths[generateKey(supportedQuery, returnType)] = new Function(
        'contextItem',
        'domFacade',
        'runtimeLibrary',
        jsResult.code
      ) as () => {};
    }
  }

  compileIfPossible(supportedQuery, ReturnType.BOOLEAN);
  compileIfPossible(unsupportedQuery, ReturnType.BOOLEAN);

  // Run time.
  function executeQuery(query: string, returnType: ReturnType) {
    const precompiledQuery = precompiledXPaths[generateKey(query, returnType)];
    if (precompiledQuery) {
      chai.assert.isTrue(executeCompiledXPath(precompiledQuery, document));
    } else {
      // Use expression backend.
      chai.assert.isTrue(evaluateXPathToBoolean(query, document));
    }
  }

  executeQuery(supportedQuery, ReturnType.BOOLEAN);
  executeQuery(unsupportedQuery, ReturnType.BOOLEAN);
}
```
I see one issue with avoiding `new Function` in FontoXPath regarding the 'auto'/fallback option. Somehow FontoXPath has to be able to use `new Function` on its own when using 'auto'. Maybe the right way to do this would be to `fontoxpath.enableFunction(Function);`. The advantage of this would be that the user does not have to know about the arguments a executable XPath as JavaScript function can receive. The downside is that use of `new Function` is not transparent at all. This solution looks kind of weird as well.

Maybe, if there is no better way, the 'auto' and 'js-codegen' options should be removed and be replaced with the `compileXPathToJavaScript` and `executeCompiledXPath` API's where the user is responsible. That way there are no 'hidden' uses of `new Function` and is using js-codegen very much opt-in. The downside is that more responsibility is shifted to the user, who has to know more about FontoXPath's internals, which can restrict future API changes. This would, however, be a good thing for security, since all use of `new Function` is explicit. This does not take away the problem that generated JavaScript could contain malicious code which is evaluated anyway... 

What do you think?